source/proxy_strategy.c

```diff
@@ -1105,6 +1548,336 @@ struct aws_http_proxy_strategy_factory *aws_http_proxy_strategy_factory_new_tunn
+    struct aws_allocator *allocator,
+    struct aws_http_message *request,
+    struct aws_string *credential_response) {
+
+    struct aws_byte_buf base64_input_value;
+    AWS_ZERO_STRUCT(base64_input_value);
+
+    struct aws_byte_buf header_value;
+    AWS_ZERO_STRUCT(header_value);
+
+    int result = AWS_OP_ERR;
+
+    if (aws_byte_buf_init(&base64_input_value, allocator, credential_response->len + 1)) {
+        goto done;
+    }
+
+    /* First build a buffer with "ntlm credential/response" in it */
+    struct aws_byte_cursor credential_response_cursor = aws_byte_cursor_from_string(credential_response);
+    if (aws_byte_buf_append(&base64_input_value, &credential_response_cursor)) {
+        goto done;
+    }
+
+    struct aws_byte_cursor base64_source_cursor =
+        aws_byte_cursor_from_array(base64_input_value.buffer, base64_input_value.len);
+
+    /* Figure out how much room we need in our final header value buffer */
+    size_t required_size = 0;
+    if (aws_base64_compute_encoded_len(base64_source_cursor.len, &required_size)) {
+        goto done;
+    }
+
+    required_size += s_proxy_authorization_header_ntlm_prefix->len + 1;
+    if (aws_byte_buf_init(&header_value, allocator, required_size)) {
+        goto done;
+    }
+
+    /* Build the final header value by appending the authorization type and the base64 encoding string together */
+    struct aws_byte_cursor ntlm_prefix = aws_byte_cursor_from_string(s_proxy_authorization_header_ntlm_prefix);
+    if (aws_byte_buf_append_dynamic(&header_value, &ntlm_prefix)) {
+        goto done;
+    }
+
+     /*
+     * We do not need base64 encoding as the credential or response are already base64 encoded
+     */
+
+    uint8_t length = (uint8_t)s_proxy_authorization_header_ntlm_prefix->len;
+
+    /* copy the length of credential into header value buffer starting with an offset*/
+    for (uint16_t i = 0; i < base64_input_value.len; i++) {
+
+        header_value.buffer[i + length] = base64_input_value.buffer[i];
+    }
+
+    header_value.len += base64_input_value.len;
+    
+    struct aws_http_header header = {
+        .name = aws_byte_cursor_from_string(s_proxy_authorization_header_name),
+        .value = aws_byte_cursor_from_array(header_value.buffer, header_value.len),
+    };
+
+    if (aws_http_message_add_header(request, header)) {
+        goto done;
+    }
+
+    result = AWS_OP_SUCCESS;
+
+done:
+
+    aws_byte_buf_clean_up(&header_value);
+    aws_byte_buf_clean_up(&base64_input_value);
+
+    return result;
+}
+
+static void s_ntlm_tunnel_transform_connect(
+    struct aws_http_proxy_strategy *proxy_strategy,
+    struct aws_http_message *message,
+    aws_http_proxy_strategy_terminate_fn *strategy_termination_callback,
+    aws_http_proxy_strategy_http_request_forward_fn *strategy_http_request_forward_callback,
+    void *internal_proxy_user_data) {
+
+    struct aws_http_proxy_strategy_tunneling_chain *ntlm_strategy = proxy_strategy->impl;
```

Comment:
shouldn't be a chain

Reply:
Fixed, all strategy are now getting casted to correct structures